apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -184,3 +256,87 @@ func validateImageURL(imageURL string) error {
 
 	return nil
 }
+
+func checkStatusAnnotation(bmhstatus *BareMetalHostStatus) error {
+
+	operatstatus_allowed := []string{"", "OK", "discovered", "error", "delayed", "detached"}
+	errortype_allowed := []string{"", "provisioned registration error", "registration error", "inspection error", "preparation error", "provisioning error", "power management error"}
+
+	if !stringInSlice(operatstatus_allowed, string(bmhstatus.OperationalStatus)) {
+		return fmt.Errorf("invalid OperationalStatus in StatusAnnotation")
+	}
+
+	if !stringInSlice(errortype_allowed, string(bmhstatus.ErrorType)) {
+		return fmt.Errorf("invalid ErrorType in StatusAnnotation")
+	}
+
+	return nil
+}
+
+func validHwdDetailsAnnotation(hwdetannotaiton string, inspect string) error {
+	if hwdetannotaiton == "" {
+		return nil
+	}
+
+	if inspect != "disabled" {
+		return fmt.Errorf("inspection has to be disabled for HardwareDetailsAnnotation, check if {'inspect.metal3.io' : 'disabled'}")
+	}
+
+	objStatus := &HardwareDetails{}
+	err := json.Unmarshal([]byte(hwdetannotaiton), objStatus)
+	if err != nil {
+		return fmt.Errorf("failed to fetch Hardware Details from annotation")
+	}
+
+	deco := json.NewDecoder(strings.NewReader(hwdetannotaiton))
+	deco.DisallowUnknownFields()
+	if err = deco.Decode(objStatus); err != nil {
+		return fmt.Errorf("invalid field in Hardware Details Annotation")
+	}
+
+	return nil
+}
+
+func validInspectAnnotation(inspectannotation string) error {
+
+	if !stringInSlice([]string{"disabled", ""}, inspectannotation) {
+		return fmt.Errorf("invalid value for Inspect Annotation")
+	}
+
+	return nil
+}
+
+func validRebootAnnotation(rebootannotation string) error {
+	if rebootannotation == "" {
+		return nil
+	}
+
+	objStatus := &RebootAnnotationArguments{}
+	err := json.Unmarshal([]byte(rebootannotation), objStatus)
```

Comment:
Is it actually a JSON? I would expect it to be a normal string.

Reply:
We have these 2 types - `reboot.metal3.io: ""` and `reboot.metal3.io/my-unique-key: "{'mode':'soft'}"`, and JSON takes care of both of these.